File: ckgame-cpp/src/core/traits/traits.cpp
#include "core/traits/traits.h"

namespace ckgame {
namespace traits {

const std::vector<Trait>& All() {
    static const std::vector<Trait> kTraits = {
        {"brave", "勇敢", Kind::Personality},
        {"craven", "怯懦", Kind::Personality},
        {"just", "公正", Kind::Personality},
        {"arbitrary", "任性", Kind::Personality},
        {"ambitious", "野心勃勃", Kind::Personality},
        {"content", "安于现状", Kind::Personality},
        {"temperate", "节制", Kind::Personality},
        {"gluttonous", "暴食", Kind::Personality},
        {"charitable", "仁慈", Kind::Personality},
        {"greedy", "贪婪", Kind::Personality},
        {"diligent", "勤奋", Kind::Personality},
        {"lazy", "懒惰", Kind::Personality},
        {"patient", "耐心", Kind::Personality},
        {"wroth", "暴怒", Kind::Personality},
        {"humble", "谦逊", Kind::Personality},
        {"proud", "傲慢", Kind::Personality},
        {"honest", "诚实", Kind::Personality},
        {"deceitful", "狡诈", Kind::Personality},
        {"gregarious", "合群", Kind::Personality},
        {"shy", "害羞", Kind::Personality},
        {"zealous", "狂热", Kind::Personality},
        {"cynical", "愤世嫉俗", Kind::Personality},
        {"strong", "强壮", Kind::Physical},
        {"weak", "虚弱", Kind::Physical},
        {"beautiful", "貌美", Kind::Physical},
        {"scarred", "有伤疤", Kind::Physical},
        {"brilliant", "聪慧", Kind::Physical},
        {"slow", "迟钝", Kind::Physical},
    };
    return kTraits;
}

const Trait* Find(const std::string& id) {
    for (const auto& t : All()) {
        if (t.id == id) return &t;
    }
    return nullptr;
}

}  // namespace traits
}  // namespace ckgame

File: ckgame-rust/src/core/traits.rs
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Kind {
    Personality,
    Physical,
    Lifestyle,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Trait {
    pub id: String,
    pub name: String,
    pub kind: Kind,
}

pub fn find(id: &str) -> Option<&'static Trait> {
    ALL_TRAITS.iter().find(|t| t.id == id)
}

pub fn all() -> &'static [Trait] {
    ALL_TRAITS
}

const ALL_TRAITS: &[Trait] = &[
    Trait { id: "brave".to_string(), name: "勇敢".to_string(), kind: Kind::Personality },
    Trait { id: "craven".to_string(), name: "怯懦".to_string(), kind: Kind::Personality },
    Trait { id: "just".to_string(), name: "公正".to_string(), kind: Kind::Personality },
    Trait { id: "arbitrary".to_string(), name: "任性".to_string(), kind: Kind::Personality },
    Trait { id: "ambitious".to_string(), name: "野心勃勃".to_string(), kind: Kind::Personality },
    Trait { id: "content".to_string(), name: "安于现状".to_string(), kind: Kind::Personality },
    Trait { id: "temperate".to_string(), name: "节制".to_string(), kind: Kind::Personality },
    Trait { id: "gluttonous".to_string(), name: "暴食".to_string(), kind: Kind::Personality },
    Trait { id: "charitable".to_string(), name: "仁慈".to_string(), kind: Kind::Personality },
    Trait { id: "greedy".to_string(), name: "贪婪".to_string(), kind: Kind::Personality },
    Trait { id: "diligent".to_string(), name: "勤奋".to_string(), kind: Kind::Personality },
    Trait { id: "lazy".to_string(), name: "懒惰".to_string(), kind: Kind::Personality },
    Trait { id: "patient".to_string(), name: "耐心".to_string(), kind: Kind::Personality },
    Trait { id: "wroth".to_string(), name: "暴怒".to_string(), kind: Kind::Personality },
    Trait { id: "humble".to_string(), name: "谦逊".to_string(), kind: Kind::Personality },
    Trait { id: "proud".to_string(), name: "傲慢".to_string(), kind: Kind::Personality },
    Trait { id: "honest".to_string(), name: "诚实".to_string(), kind: Kind::Personality },
    Trait { id: "deceitful".to_string(), name: "狡诈".to_string(), kind: Kind::Personality },
    Trait { id: "gregarious".to_string(), name: "合群".to_string(), kind: Kind::Personality },
    Trait { id: "shy".to_string(), name: "害羞".to_string(), kind: Kind::Personality },
    Trait { id: "zealous".to_string(), name: "狂热".to_string(), kind: Kind::Personality },
    Trait { id: "cynical".to_string(), name: "愤世嫉俗".to_string(), kind: Kind::Personality },
    Trait { id: "strong".to_string(), name: "强壮".to_string(), kind: Kind::Physical },
    Trait { id: "weak".to_string(), name: "虚弱".to_string(), kind: Kind::Physical },
    Trait { id: "beautiful".to_string(), name: "貌美".to_string(), kind: Kind::Physical },
    Trait { id: "scarred".to_string(), name: "有伤疤".to_string(), kind: Kind::Physical },
    Trait { id: "brilliant".to_string(), name: "聪慧".to_string(), kind: Kind::Physical },
    Trait { id: "slow".to_string(), name: "迟钝".to_string(), kind: Kind::Physical },
];